### ISSUE TITLE
`CurrentRow` is not fit for purpose

### ISSUE THREAD

**Issue by jamesrhester** [COMCIFS/dREL#6]
dREL operates in a file-free, block-free space and therefore is not aware of whether or not blocks have been merged. If multiple space groups are present (for example, a powder diffraction dataset containing several structures) `space_group_symop` will contain several lists of symops, which must each be numbered in the sequence expected if they were in individual blocks. `CurrentRow` will not do this.

I suggest that it is redefined as `CurrentRow(object_id)`, where it enumerates the rows by first filtering the category into rows that have particular values for all key data names except for `object_id`. In the case of `space_group_symop`, this would be `CurrentRow(id)`. If no argument is provided, the current behaviour would be preserved.

This affects only the `space_group_symop.id` data name, and software that has implemented `CurrentRow`.

**Comment by vaitkus** [comment]
> I suggest that it is redefined as `CurrentRow(object_id)`, where it enumerates the rows by first filtering the category into rows that have particular values for all key data names except for `object_id`. In the case of `space_group_symop`, this would be `CurrentRow(id)`. If no argument is provided, the current behaviour would be preserved.

This seems a bit strange to me, since in order to assign the `id` value, the  `CurrentRow(id)` method is called which explicitly ignores the `id` column (that is, we ignore values that we are trying to generate and are thus not yet there). Wouldn't it be clearer to provide `CurrentRow` with one or more `object_id`s that should serve as the key when determining the sequential numeric values? For example, if we assume that the `_space_group_symop.set_id` data item identifies different symop lists, the appropriate call would then be `CurrentRow(set_id)`. What do you think?

Finally, I was not able to find the description of the `CurrentRow()` neither in this repository, nor in the supplementary material of the original dREL paper. Regardless of which approach is chosen for `CurrentRow()`, I think it would be a great time to explicitly describe it in this repo.

**Comment by jamesrhester** [comment]
Yes, `CurrentRow` was found only in this one method in the original DDLm dictionaries, and was nowhere documented. It is documented in the forthcoming chapter of Vol G as one of the built-in functions, but I agree it would be worth specifying all of those built-in functions in this repository.

The problem with `CurrentRow(set_id)` is that arbitrary further key data names might be added during future expansion, and then the argument list would need to be expanded. But of course you are right that there are no values of `_space_group_symop.id` to count, so that won't work. So how about `CurrentRow(operation_xyz)`? So the revised description would be:

`CurrentRow(object_id)`: return the current row in the category, after filtering the category on the current values for all key data names, excluding from the list of key data names `object_id` and the present data name, if they are key data names. If no argument is provided, the category is filtered on the current values for all key data names, except the present data name.

I'll note that this function should only ever be used in this particular dREL method, to support legacy symop constructions.

**Comment by vaitkus** [comment]
> `CurrentRow(object_id)`: return the current row in the category, after filtering the category on the current values for all key data names, excluding from the list of key data names `object_id` and the present data name, if they are key data names. If no argument is provided, the category is filtered on the current values for all key data names, except the present data name.

If I correctly understand the updated definition, an explicit parameter is no longer even required at all. That is, the `CurrentRow(operation_xyz)` call is equivalent to the `CurrentRow()` call since `operation_xyz` is not a key data name. The  `CurrentRow()` call with no parameters would also work  inside the definition of `_space_group_symop.id` since the "present item" (`id`) would be automatically removed and would thus function properly with merged datasets.

I would, however, like to suggest a slight rephrasing of the definition:
**CurrentRow()**: _Returns the index of the current row in the category. The row index is assigned by first grouping all rows based on the category key data items and then assigning sequential integers to the rows with each group having its own 1-based index sequence. The present data item must not be used in the grouping even if it defined as a category key data item._

I did not include the part about calling the function with a parameter as it may not be required at all, but it can be added at the end of the definition later on if needed. Also, I am not completely sure if the term "present data item" is clear enough so we might need to define it somewhere in the document as well. Finally, once the definition is finalised, a few complementing examples would go a long way of explicitly describing the behaviour of this function.

However, after thinking more deeply on the whole merged data set situation, I have a hunch that the proposed enhancements to the `CurrentRow()` function might find little applicability in real life. If one of the main goals is to enhance compatibility with legacy files, the handling of merged datasets is a bit of an overkill since legacy files are almost guaranteed to not contain `_space_group_symop.set_id` or similar key items needed to merge the datasets.  Of course, these new key items might also contain dREL methods that assign a unique id to each data block, but this seems like a very exotic scenario. Or is there a more common case that I might be missing?

Finally, I noticed that `CurrentRow()` seems to be spelt as `Current_Row()` in the definition of `_space_group_symop.id` so we should eventually choose one of the two names.

**Comment by jamesrhester** [comment]
I like your rephrased definition.

In terms of potential scenarios, legacy powder diffraction files contain multiple structures spread over several data blocks. The modern DDLm dictionaries explain how these (even legacy) data blocks can be merged into a notional single collection of tables, so it is entirely possible that part of this merging would involve generation of `_space_group_symop.id` values for more than one space group for legacy files (and probably files that are being generated even today, I'll have to check, software changes very slowly). So I think it is safest to keep the detailed definition that you have provided.

Yes, it should be `Current_Row`.

**Comment by jamesrhester** [comment]
Note also that `Current_Row` counts from 0, in keeping with dREL's python heritage.

**Comment by jamesrhester** [comment]
Reopening this: implementing the new `CurrentRow` function throws up a subtle-ish problem. Normally, the form `function(variable)` means either "call `function` with the value of `variable`" or in pass-by-reference, "call `function` with the address of `variable`". Our interpretation above is neither of these: we are calling `function` with the _name_ of `variable` in order to exclude that variable from grouping.

So I'd like to suggest that the argument to `CurrentRow` should be a string, i.e `CurrentRow("cat.obj")`.

It is frustrating that we have to spend so much time on a function that only exists for legacy reasons.

**Comment by vaitkus** [comment]
I think I see where you are coming from. Would the same logic also apply to  `Unique_id(<cat>.<obj>)`?

Also, I think we call the function is called `Current_row()` not `CurrentRow()` (according to https://github.com/COMCIFS/dREL/blob/main/builtins.md), but I guess it can still be renamed if that is more convenient.

**Comment by jamesrhester** [comment]
Yes, you are right that I got the name wrong, and that this applies to Unique_id as well. I'll make a PR.
